torch/_inductor/memory.py

```diff
@@ -0,0 +1,555 @@
+    for buf_name, buf_info in name_to_buf_info.items():
+        memory[buf_info.start_time] += buf_info.size_alloc
+        memory[buf_info.end_time + 1] -= buf_info.size_free
+
+    # get peak memory by compute the cumulative memories
+    max_memory = 0
+    cur_memory = 0
+    memories_at_nodes = []
+    for t in range(len(nodes) + 1):
+        cur_memory += memory[t]
+        memories_at_nodes.append(cur_memory)
+        max_memory = max(max_memory, cur_memory)
+
+    return (max_memory, memories_at_nodes)
+
+
+def assign_predcessor_and_successor_nodes_to_nodes(
+    nodes: List[BaseSchedulerNode], name_to_fused_node: Dict[str, BaseSchedulerNode]
+) -> None:
+    """
+    Assign to each scheduler node its predecessor and successor nodes.
+    """
+    from .scheduler import SchedulerBuffer
+
+    for node in nodes:
+        node.mpi_node.pred_nodes = list(
+            {
+                name_to_fused_node[pred_buffer.defining_op.get_name()]
+                for pred_buffer in node.mpi_node.pred_buffers
+                if (
+                    isinstance(pred_buffer, SchedulerBuffer)
+                    and pred_buffer.defining_op.get_name() in name_to_fused_node
+                )
+            }
+        )
+        node.mpi_node.succ_nodes = list(
+            {
+                succ_node
+                for buffer in node.get_outputs()
+                for succ_node in buffer.mpi_buffer.succ_nodes
+            }
+        )
+
+
+def topological_sort_lpmf(
+    nodes: List[BaseSchedulerNode],
+    name_to_input_buf: Dict[str, FreeableInputBuffer],
+    name_to_buf: Dict[str, SchedulerBuffer],
+    graph_outputs: Set[str],
+) -> List[BaseSchedulerNode]:
+    """
+    A bfs-based greedy topological order. LPMF stands for "Least Peak Memory First".
+    The algorithm maintain the max memory so far.
+    At every iteration, for each scheduleable node, it computes:
+        - how much memory needs to be allocated for the output buffers of this node;
+        - how much memory can be freed as a result of executing this node.
+    This gives us two values for each node:
+        (1) mem1: memory during the execution of the node;
+        (2) mem2: memory after executing the node, after some input buffers are freed.
+    The greedy approach select as follows:
+        (i) if there are nodes whose mem1 values are below the max memory so far,
+            then pick the node with the lowest mem2 value;
+        (ii) otherwise, pick the one with the lowest mem1 value.
+    """
+
+    # compute nodes' number of unmet dependencies (for schedulability)
+    # initialize the list of nodes ready to be scheduled
+    nodes_to_schedule: Set[BaseSchedulerNode] = set()
+    for node in nodes:
+        # note that .unmet_dependencies could have deps with the same name
+        # and in that case, it should only be counted once
+        node.mpi_node.indegree = len(node.mpi_node.pred_nodes)
+        if node.mpi_node.indegree == 0:
```

Comment:
Like other comments, if we are going to inplace update the class, it would have the not yet computed state be explicitly represented (None), instead of a value which is a valid state later. It makes it more clear as we are following the logic / debugging.